spec/API_specification/array_object.md

```diff
@@ -366,6 +366,23 @@ Evaluates `x1_i & x2_i` for each element `x1_i` of an array instance `x1` with t
 Element-wise results must equal the results returned by the equivalent element-wise function [`bitwise_and(x1, x2)`](elementwise_functions.md#logical_andx1-x2-).
 ```
 
+(method-__dlpack__)=
+### \_\_dlpack\_\_(/, *, stream=None)
+
+Exports the array as a DLPack capsule, for consumption by {ref}`function-from_dlpack`.
+
+#### Parameters
+
+-   **stream**: _Optional\[int\]_
+
+    -   If given, the CUDA or ROCm stream number the consumer will use. Default is `None`, which means the legacy default stream.
```

Comment:
Note that this is different from how `stream` is specified in https://numba.readthedocs.io/en/latest/cuda/cuda_array_interface.html#python-interface-specification. I actually don't understand that spec - it says for `None` that no synchronization is needed, it uses 1/2 for legacy/per-thread default stream and other integers for non-default streams. Which seems odd - what if the stream number of a non-default stream in use is `2` for example?

Using:
- `None`: legacy default stream
- `0`: per-thread default stream
- `1, 2, ...` non-default stream numbers
seems to make more sense. @leofang am I missing something there?

Reply:
@rgommers `None` is actually confusing across libraries. For example, in Numba `None` means there is no Numba's default stream (to be distinguished from "CUDA's (whichever) default stream"), whereas in CuPy `None` simply refers to CUDA's default stream, which in turn is `1` (the legacy default stream), though we're on the process of adopting `2` (the per-thread default stream). `0` is not acceptable either for the same reason: it's semantically unclear depending on [how the libraries containing CUDA code are compiled](https://docs.nvidia.com/cuda/cuda-runtime-api/stream-sync-behavior.html#stream-sync-behavior) and the runtime behavior defined in the Python hooks.

Note that in CUDA you don't get to choose the stream numbers --- CUDA macro-defines [`1` for `cudaStreamLegacy`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1g4e62d09dde16ba457b0a97f3a5262246) and [`2` for `cudaStreamPerThread`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1g7b7129befd6f52708309acafd1c46197), which CAI v3 followed. Any user/non-default stream created via `cudaStreamCreate()` is guaranteed to start on or after `3`. (In fact the CUDA driver would reserve a stream pool internally, so the actual start number is way after `3`.) I hope this makes CAI v3 clearer to you.

I'll try to catch up the rest of the discussions here as well as in the DLPack repo after Monday (tomorrow)...😅

Reply:
btw fun fact: in HIP syncing over the stream `1` or `2` would lead to segfault, as HIP does not support them: https://github.com/cupy/cupy/pull/4458#issuecomment-747884690.

Reply:
Thanks for the explanation and links @leofang. I updated it to match `__cuda_array_interface__`.

Reply:
Thanks, @rgommers. Apology, I realized my first sentence wasn't complete; it should have been
> `None` is actually confusing across libraries, *so we decided to use it to mark the situation "there is no need to do any kind of synchronization for whatever reason, just take the device pointer and do your work"*.